README.md
Cinnamon Welcome
==============

Welcome screen application to greet new users on their first login.

[![Get it from the Snap Store](https://snapcraft.io/static/images/badges/en/snap-store-white.svg)](https://snapcraft.io/ubuntu-ubuntucinnamon-welcome)

### Translators

We need to translate ubuntucinnamon-welcome into your language/locale.  Please
help-out - visit:

https://www.transifex.com/ubuntu-cinnamon/ubuntucinnamon-welcome/

### Installation

```sh
sudo apt install gir1.2-webkit2-4.0 python3-notify2 node-uglify
git clone https://github.com/Ubuntu-Cinnamon-Remix/ubuntucinnamon-welcome.git

uglifyjs -o data/js/bootstrap.min.js data/js/bootstrap.js
uglifyjs -o data/js/baguetteBox.min.js data/js/baguetteBox.js
uglifyjs -o data/js/popper.min.js data/js/popper.js
```

### Testing

```sh
cd ubuntucinnamon-welcome
./ubuntucinnamon-welcome -d
```

To test the live session

    ./ubuntucinnamon-welcome --force-session=live -d
    
To test a language or locale

    ./ubuntucinnamon-welcome --locale=pl
    
change "pl" to your language i18n shortcode

### Compile scss files
```sh
 ./sassc-compile.sh
```

### Screenshot
#### Live session
![Screenshot of Cinnamon Remix Welcome App](https://raw.githubusercontent.com/cinnamon-remix/ubuntucinnamon-welcome/master/screenshot-live-session.png)

#### Normal session
![Screenshot of Cinnamon Remix Welcome App](https://raw.githubusercontent.com/cinnamon-remix/ubuntucinnamon-welcome/master/screenshot-normal-session.png)

### LICENSES
  The project contains files with a variety of licenses.
  For the complete list of licenses, see [debian/copyright](https://github.com/cinnamon-remix/ubuntucinnamon-welcome/blob/master/debian/copyright) file.


debian/changelog
ubuntucinnamon-welcome (0.0.1) focal; urgency=medium

  * Initial fork
  * Changed budgie colors to #DD682A (cinnamon) colors
  * Converted all "budgie" to "cinnamon"
  * Removed Discourse
  * Replaced IRC with Telegram
  * Update icons
  * Replace all Ubuntu Budgie with Ubuntu Cinnamon
  * I can't think of anything else.

 -- Joshua Peisach <ubuntucinnamonremix7@gmail.com>  Wed, 29 Jan 2020 19:20:15 -0500

budgie-welcome (0.6.0) bionic; urgency=medium

  * Bug-fix to display live booting information correctly in tabs
  * Remove custom CSS for adapta theme
  * Adapta theme now installs breeze cursor and papirus icons
  * latest translations
  * Fix crash when displaying version from command line (LP: #1721711)
  * Fix crash when graphics info cannot be found (LP: #1720623)
  * Add menu accessible .desktop files for themes and applets
  * Rename flat-plat to materia theme
  * Add cinnamon-extras applets
  * Add Ant theme and Pop theme
  * Add links to project websites for applets
  * Add recommendation for Dropbox
  * Drop recommendation for cinnamon-screenshot-applet
  * Headerbar GTK theming
  * Packaging Changes:
    - debian/control Bump standard version: No changes required
    - debian/control debhelper v11
    - debian/control corrected Vcs locations
    - debian/rules remove unneeded autoinstall override
    - debian/copyright: 2018 year change

 -- David Mohammed <fossfreedom@ubuntu.com>  Mon, 22 Jan 2018 20:16:38 +0000

budgie-welcome (0.5.5) artful; urgency=medium

  * Bug-fix release
    - install missing package for flatpak install option
      (LP: #1719251)
    - Latest translation updates

 -- David Mohammed <fossfreedom@ubuntu.com>  Tue, 26 Sep 2017 20:52:01 +0100

budgie-welcome (0.5.4) artful; urgency=medium

  * Bug-fix release (LP: #1717791)
    - hide cinnamon-pixel-saver-applet and cinnamon-appmenu-applet
      since these do not work with cinnamon desktop 10.4 (16.04)
    - changed flatpak weblink to new flathub equivalent
    - Latest translation updates

 -- David Mohammed <fossfreedom@ubuntu.com>  Sun, 17 Sep 2017 23:39:51 -0500

budgie-welcome (0.5.3) artful; urgency=medium

  * bug-fix release (LP: #1712047)
    - rework to ensure GTK+ calls are made on the main thread
      to avoid segmentation faults when webkit2gtk version
      >= 2.17.90

 -- David Mohammed <fossfreedom@ubuntu.com>  Thu, 07 Sep 2017 16:39:29 -0500

budgie-welcome (0.5.2) artful; urgency=medium

  * Bug-fix release (LP: #1713210)
    - temporary patch source to workaround showstopper segmentation faults
      caused by experimental webkit2gtk library upload in artful
      (upstream webkit2gtk issue 175855)
    - update VCS fields in debian/control to reflect the branched code
    - temporarily disable optional wmmover-applet since not available to
      be installed yet
    - update cinnamon-shop picture to cover missing desktop purchase options
    - clean up source tree so that locale binaries are not included with
      the source; these are created as part of the build
  * Latest translation updates

 -- David Mohammed <fossfreedom@ubuntu.com>  Fri, 25 Aug 2017 17:31:34 +0100

budgie-welcome (0.5.1) artful; urgency=medium

  * New Release (LP: #1711912)
  * Add installation options for cinnamon applets:
    Global Menu,
    WMMover,
    SysMonitor
  * Updated Translations

 -- David Mohammed <fossfreedom@ubuntu.com>  Thu, 03 Aug 2017 18:42:51 +0100

budgie-welcome (0.5.0) artful; urgency=medium

  * New Release (LP: #1701120)
  * Add custom window decoration to allow app move via mouse
  * Translation Support
  * Online Shop replaces Community Chatroom
  * Install Software button on front-page to allow installation
    for flatpak, snap and new cinnamon-applet page
  * Cinnamon Applet page allows curated list of Cinnamon Desktop applets
    to be installed
  * Themes revamp including lightbox support
  * Default Settings reworked to differentate between makeover theming,
    reset defaults and layout manager
  * Uplift to bootstrap v4 alpha 6
  * ibus/fcitx switch
  * Packaging Changes
    - debian/copyright - http-->https
      updates to bootstrap copyright year

 -- David Mohammed <fossfreedom@ubuntu.com>  Sat, 03 Jun 2017 08:59:08 +0100

budgie-welcome (0.4.5) zesty; urgency=medium

  * bug-fix release (LP: #1669589)
    - replace shortcut key for skippyxd to prevent conflict with chromium
    - inform users to reboot after installing flatpak
    - copyright year and name update for source file budgie-welcome
    - make sure chrome only installable for 64bit installs
    - ensure start page elements align correctly for Chinese locale
    - fix package and ppa installs so that material theme is correctly
      installed
    - Update Material designs with missing Noto fonts
    - budgie-welcome source trailing spaces removed
    - recommendations has backports now listed first rather than last

 -- David Mohammed <fossfreedom@ubuntu.com>  Fri, 24 Feb 2017 19:16:55 +0000

budgie-welcome (0.4.4) zesty; urgency=medium

  * bug-fix release (LP: #1665923)

scripts/welcome-po.py
#! /usr/bin/python3
# -*- coding:utf-8 -*-
#
# Copyright (C) 2016 Robin Thompson (robint99)
#
# welcome-po is free software: you can redistribute it and/or modify
# it under the temms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# welcome-po is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with welcome-po. If not, see <http://www.gnu.org/licenses/>.
#
# Originally written for ubuntucinnamon-welcome. Adapted for ubuntucinnamon-welcome.
#


"""  i18n helper for Cinnamon Welcome

Perform one of the following functions depending on the command line args

    create a .pot file for the Welcome program and put it in the ./po directory
    (create this if necessary). To be used when initially creating the .pot file
    and also when new strings to translate are included in the Welcome program

    create a .po file either for a specified locale or for all locales supported
    by the system. If a .po file already exists, it will not be overwritten since
    it may contain translations

    update all .po files with new translatable strings from the .pot file

    compile and install the translations in all available .po file to
    usr/share/locale

"""

import os
import sys
import subprocess
import shutil
import glob

###########################################################
def show_usage():
    """ Display the command line options """

    print("\nwelcome-po usage")
    print("\nUsage: welcome-po [arguments]")
    print("  --create-pot                Create a ubuntucinnamon-welcome.pot in the po directory")
    print("                              (can also be used when whenever new translatable strings")
    print("                               are added to ubuntucinnamon-welcome.\n")
    print("  --update-pos                Update all .po files in the .po directory with new")
    print("                              translateable strings from the .pot file\n")
    print("  --install                   Compile all of the .po files in the po directory")
    print("                              and install them under ./locale/\n")
    print("  --help                      Show this message\n")
    print(" Requirements:                Must be run in the same directory as ubuntucinnamon-welcome")
    print("                              Requires xgettext, pot2po, msgmerge, msgfmt \n")



###########################################################
def create_pot(po_dir):
    """ Create .pot file for Ubuntu Mate Welcome

    Expect to find Welcome in the current directory...
    Create a po directory if one doesn't exist

    Args:
        po_dir - the location of the po directory

    """

    if not subprocess.call(["which", "pygettext"],
                       stdout=subprocess.PIPE, stderr=subprocess.PIPE) == 0:
        print("Error: pygettext is not available.")
        sys.exit(1)

    if not (os.path.exists(po_dir)):
        os.mkdir(po_dir)

    pot_file = os.path.join(po_dir,"ubuntucinnamon-welcome.pot")

    subprocess.call(["xgettext",
	             "-d", "ubuntucinnamon-welcome",    # domain name
	             "-o", pot_file,                 # output file
                     "ubuntucinnamon-welcome",          # input file
                     "-L", "Python"])                # Language

    print ("%s created.\n" %pot_file)


###########################################################
def update_pos(po_dir):
    """ Update all .po files with any new translatable strings from the .pot
        file

    Use msgmerge in order to preserve any existing translations in the .po
    files

    Args: po_dir - the directory containing the .po files
    """

    pot_file = os.path.join(po_dir,"ubuntucinnamon-welcome.pot")
    if not os.path.exists(pot_file):
        print("Error: ubuntucinnamon-welcome.pot does not exist")
        exit()

    po_files = glob.glob(os.path.join(po_dir, '*.po'))

    for po_file in po_files:
        subprocess.call(["msgmerge",
                        po_file, pot_file,
                        "-U",                # Update po file
                        "-q"])               # Quiet mode
        print("%s updated " %po_file)

    print ("Update completed......")


###########################################################
def compile_and_install(po_dir, locale_dir):
    """ Compile all of the po files in the po directory into the
        locale directory, ready for use by the Welcome
        program

    Create the locale directory and any required subdirectories
    if they don't already exist
    """

    if not (os.path.exists(locale_dir)):
        os.mkdir(locale_dir)

    po_files = glob.glob(os.path.join(po_dir, '*.po'))
    for po_file in po_files:
        locale_name =(os.path.splitext(os.path.split(po_file)[1])[0])

        #create a directory for the locale if we don't already have one
        this_locale = os.path.join(locale_dir, locale_name)
        lcm_dir = os.path.join(this_locale, "LC_MESSAGES")

        if not os.path.exists(this_locale):
            os.mkdir(this_locale)
            # also make the LC_MESSAGES directory
            os.mkdir(lcm_dir)

        #now compile and install the .po
        print ("processing %s" %po_file)
        output_file = os.path.join(lcm_dir, "ubuntucinnamon-welcome.mo")

        subprocess.call(["msgfmt", po_file,
	             "--output-file", output_file])


    print ("All languages compiled.")


if (len(sys.argv)==1) or (sys.argv[1]=="--help"):
    show_usage()
    exit()

if not os.path.exists("../ubuntucinnamon-welcome"):
    print("Error: Need to be in the scripts directory for ubuntucinnamon-welcome...")
    exit()

source_dir = '../'
po_dir = os.path.join(source_dir, "po")
locale_dir = os.path.join(source_dir, "locale")

arg = sys.argv[1]
if (arg == "--create-pot") or (len(sys.argv)==1):
    create_pot(po_dir)
    exit()

if arg=="--update-pos":
    update_pos(po_dir)
    exit()

if arg=="--install":
    compile_and_install(po_dir, locale_dir)
    exit()



data/js/welcome.js
// Which page?
current_page = document.location.href.match(/[^\/]+$/)[0];

// Pass commands to Python
function cmd(instruction) {
    document.title = instruction;
}

// Global across all pages
$(window).load(function () {
    // Smoothly fade into the page.
    $('body').fadeIn('slow');
});

// Smoothly fade out of the page.
function smoothPageFade(target_href) {
    $('html').fadeOut('fast');
    setTimeout(function () {
        window.location.href = target_href;
    }, 200);
}

// When page first opens
$(document).ready(function () {

    // Back to the top
    $(window).scroll(function () {
        if ($(this).scrollTop() > 90) {
            $('#scroll-top').fadeIn();
        } else {
            $('#scroll-top').fadeOut();
        }
    });

    $('#navigation-right').append('<a id="scroll-top" class="navigation-button" style="display:none"><i class="material-icons">&#xE316;</i></a>')

    $('#scroll-top').click(function () {
        $("html, body").animate({
            scrollTop: 0
        }, 600);
        return false;
    });

});

// Smoothly fade between two elements (by ID)
function smoothFade(from, to) {
    $(from).fadeOut();
    setTimeout(function () {
        $(to).fadeIn();
    }, 400);
}

// Smoothly fade the navigation sub-title
function changeSubtitle(textToDisplay) {
    // Smoothly fade subtitle
    $('#navigation-sub-title').fadeOut();
    setTimeout(function () {
        $('#navigation-sub-title').html(textToDisplay);
        $('#navigation-sub-title').fadeIn();
    }, 400);
}

// Remove slowly, fadeOut an element and then remove it.
function removeSlowly(element) {
    $(element).fadeOut()
    setTimeout(function () {
        $(element).remove();
    }, 1000);
}

// For pages that depend on an internet connection, but Welcome couldn't connect.
function reconnectTimeout() {
    if (!$('#reconnectFailed').is(':visible')) {
        $('#reconnectFailed').fadeIn();
    } else {
        $('#reconnectFailed').jAnimateOnce('pulse');
    }
}

// Dynamically set the cursor,
function setCursorBusy() {
    $('html').addClass('cursor-wait');
    $('body').addClass('cursor-wait');
    $('a').addClass('cursor-wait');
}

function setCursorNormal() {
    $('html').removeClass('cursor-wait');
    $('body').removeClass('cursor-wait');
    $('a').removeClass('cursor-wait');
}

///////////////////////////////////////////////////////////////

// Main Menu Only = Animation
if (current_page == 'index.html') {

    // Animate elements of the page
    $('#main-menu-logo').jAnimateOnce('rotateIn');
    $('.main-menu-text').fadeIn('slow');
    $('#open-at-start').jAnimateOnce('fadeIn');
    setTimeout(function () {
        $('#mate-blur').jAnimateOnce('zoomIn');
        $('#mate-blur').show();
    }, 50);

    // Have we greeted the user already?
    if (document.cookie == 'greeted=yes') {

        $(document).ready(function () {
            $(".fade").removeClass("fade");
            $(".fade-1s").removeClass("fade-1s");
            $(".fade-2s").removeClass("fade-2s");
            $(".fade-3s").removeClass("fade-3s");
            $(".fade-4s").removeClass("fade-4s");
            $(".fade-5s").removeClass("fade-5s");
        });
    }

    document.cookie = 'greeted=yes';

    // Enable tooltips
    /* $(document).ready(function() {
       $("body").tooltip({ selector: '[data-toggle=tooltip]' });
     });*/

    // Sssh... You found the little secrets! ;)
    //// Logo starts to animate after a minute.
    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('tada');
    }, 60000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('flip');
    }, 60000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('rotateOut');
    }, 70000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('rotateIn');
    }, 71000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('rollOut');
    }, 80000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('rollIn');
    }, 81000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('zoomOut');
    }, 90000);

    setTimeout(function () {
        $('#main-menu-logo').jAnimateOnce('zoomIn');
    }, 91000);

    // Internally work with days, months and years as a number.
    function dateAsNumber(day, month, year) {
        // Assumes 'month' parameter is in base 0.
        // day, month   = Required.
        // year         = Optional, set to 'null' to use current year.
        var then = new Date();
        if (year != null) {
            then.setYear(year);
        }
        then.setMonth(month - 1);
        then.setDate(day);
        var finalNumber = Math.floor(then / (1000 * 60 * 60 * 24));
        //~ var dateInYears = current + todayAsNumber;
        //~ console.log('** finalNumber: '+finalNumber)
        //~ return dateInYears;
        return finalNumber;
    }

    // Activate Confetti
    function activateConfetti() {
        if (disable_confetti == false) {
            $('#special').html('<canvas id="confetti" width="100%" height="100%" style="z-index: -1000; position: absolute; top: 0px; left: 0px;"></canvas>');
            startConfetti();
            return
        } else {
            return
        }
    }

    // Use 'dd', 'mm' and 'yyyy' variables to re-use code.
    var today = new Date();
    var dd = 0;
    var mm = 0;
    var yyyy = 0;
    var release = '';

    // What is today?
    var today = new Date();
    var todayAsNumber = dateAsNumber(today.getDate(), today.getMonth() + 1, null)

    // Check dates for special events.
    function specialEventCheck(dateNo, title_text, show_confetti, fa_icon) {
        // dateNo        = dateAsNumber(dd, mm, yyyy) function.
        // title_text    = Text to display when date matches.
        // show_confetti = True / False = Celebrate when date matches.
        // fa_icon       = FontAwesome icon to display. Usually 'bell' or 'calendar' or 'bug'.
        var do_show_this = false;
        if (dateNo == todayAsNumber) {
            // Today is the day!
            var do_show_this = true;
        }
        if (do_show_this == true) {
            $('.main-menu-text').hide();
            $('.main-menu-text').fadeIn();
            $('.main-menu-text').html("<span class='fa fa-" + fa_icon + "'></span>&nbsp;" + title_text);
            if (show_confetti == true) {
                activateConfetti();
            }
        }
    }

    // Dates to be checking for.
    function checkDates() {
        // String variables are passed via Python. See _push_config for "Special Event Strings".

        // Official Flavour Status - 26/Nov/2017
        var age = today.getFullYear() - 2017;
        dd = 26;
        mm = 11;
        yyyy = null;
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 7, flavour_anniversary_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 6, flavour_anniversary_future + ' ' + days_in + ' 6 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 5, flavour_anniversary_future + ' ' + days_in + ' 5 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 4, flavour_anniversary_future + ' ' + days_in + ' 4 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 3, flavour_anniversary_future + ' ' + days_in + ' 3 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 2, flavour_anniversary_future + ' ' + days_in + ' 2 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 1, flavour_anniversary_future + ' ' + tomorrow, true, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy), flavour_anniversary_present + ' ' + age + ' ' + years_ago, true, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 1, flavour_anniversary_past + ' ' + yesterday, true, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 2, flavour_anniversary_past + ' 2 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 3, flavour_anniversary_past + ' 3 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 4, flavour_anniversary_past + ' 4 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 5, flavour_anniversary_past + ' 5 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 6, flavour_anniversary_past + ' 6 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 7, flavour_anniversary_past + ' 7 ' + days_ago, false, 'calendar');

        // Project Birthday - 01/02/2017
        var age = today.getFullYear() - 2017;
        dd = 01;
        mm = 02;
        yyyy = null;
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 7, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 6, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 6 ' + future_days, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 5, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 5 ' + future_days, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 4, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 4 ' + future_days, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 3, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 3 ' + future_days, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 2, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + days_in + ' 2 ' + future_days, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 1, project_birthday_future + ' ' + age + ' ' + years_old + ' ' + tomorrow + ' ' + project_birthday, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy), project_birthday_present + ' ' + age + ' ' + years_old + ' ' + today_string + ' ' + project_birthday, true, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 1, project_birthday_past + ' ' + age + ' ' + years_old + ' ' + yesterday + ' ' + project_birthday, true, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 2, project_birthday_past + ' ' + age + ' ' + years_old + ', 2 ' + days_ago, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 3, project_birthday_past + ' ' + age + ' ' + years_old + ', 3 ' + days_ago, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 4, project_birthday_past + ' ' + age + ' ' + years_old + ', 4 ' + days_ago, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 5, project_birthday_past + ' ' + age + ' ' + years_old + ', 5 ' + days_ago, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 6, project_birthday_past + ' ' + age + ' ' + years_old + ', 6 ' + days_ago, false, 'birthday-cake');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 7, project_birthday_past + ' ' + age + ' ' + years_old + ', 7 ' + days_ago, false, 'calendar');

        // Holiday Celebrations
        specialEventCheck(dateAsNumber(31, 12, null), celebrate_new_year, true, 'calendar');
        specialEventCheck(dateAsNumber(01, 01, null), celebrate_new_year, true, 'calendar');

        // 18.04 Beta 2
        dd = 24;
        mm = 03;
        yyyy = 2018;
        release = 'Ubuntu MATE 18.04 Beta 2';
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 3, release + ' ' + project_release_future + ' ' + days_in + ' 3 ' + future_days, false, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 2, release + ' ' + project_release_future + ' ' + days_in + ' 2 ' + future_days, false, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 1, release + ' ' + project_release_future + ' ' + tomorrow, false, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy), release + ' ' + project_release_present, true, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 1, project_release_thanks + ' ' + release + '.', true, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 2, project_release_thanks + ' ' + release + '.', false, 'bug');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 3, project_release_thanks + ' ' + release + '.', false, 'bug');

        // 16.04 Final Release
        dd = 21;
        mm = 04;
        yyyy = 2018;
        release = 'Ubuntu Cinnamon 18.04 LTS';
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 7, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 6, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 5, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 4, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 3, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 2, release + ' ' + project_release_future + ' ' + days_in + ' 7 ' + future_days, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) - 1, release + ' ' + project_release_future + ' ' + tomorrow, true, 'bell');
        specialEventCheck(dateAsNumber(dd, mm, yyyy), release + ' ' + project_release_present, true, 'bell');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 1, release + ' ' + project_release_past + ' ' + yesterday, true, 'bell');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 2, release + ' ' + project_release_past + ' 2 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 3, release + ' ' + project_release_past + ' 3 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 4, release + ' ' + project_release_past + ' 4 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 5, release + ' ' + project_release_past + ' 5 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 6, release + ' ' + project_release_past + ' 6 ' + days_ago, false, 'calendar');
        specialEventCheck(dateAsNumber(dd, mm, yyyy) + 7, release + ' ' + project_release_past + ' 7 ' + days_ago, false, 'calendar');

        // To-do as possible improvement: Retrieve events as a list from server.
    }
}

// Getting Started Only - Index Pane for Selecting Topics
if (current_page == 'gettingstarted.html') {

    function indexOpen() {
        // Is the index already open?
        if ($('#index-menu').is(':visible')) {
            indexClose();
        } else {
            // Open the Index
            $('#index-open').addClass('disabled');
            $('#index-open').prop('disabled', true);
            $("#index-overlay").fadeIn();
            $("#index-menu").show();
            $('#index-menu').jAnimateOnce('fadeInLeft');
        }
    }

    function indexClose() {
        $('#index-open').removeClass('disabled');
        $('#index-open').prop('disabled', false);
        if ($('#index-menu').is(':visible')) {
            $("#index-overlay").fadeOut();
            $('#index-menu').jAnimateOnce('fadeOutLeft', function () {
                $("#index-menu").hide();
            });
        }
    }

    function changePage(id, humanText) {
        // 'id' is one used for <div>.
        // 'humanText' is displayed on navigation's sub title.

        indexClose();

        $('.topicContents').fadeOut();
        $('#navigation-sub-title').fadeOut();

        // Smoothly fade between topics
        setTimeout(function () {
            $('#navigation-sub-title').html(humanText);
            $('#navigation-sub-title').fadeIn();
            $('#' + id).fadeIn();
        }, 500);
    }

    // Show initial page and index pane on page load
    changePage('initial', 'Choose a Topic');
    setTimeout(function () {
        indexOpen();
    }, 500);
    $('#index-open').jAnimateOnce('fadeInDown');

    // Show additional information on the page based on checkbox state.
    $('.dualBootWin').hide();
    $('#showDualBootWin').click(function () {
        if ($(this).prop('checked') == true) {
            $('.dualBootWin').fadeIn();
        } else {
            $('.dualBootWin').fadeOut();
        }
    });

    // Graphics Detection
    // Must be executed shortly after page fully loads in order for variables to exist.
    setTimeout(function () {
        $('.graphics-pci').html(graphicsGrep);

        // Auto detection alert initially displays "failed".
        if (graphicsVendor == 'NVIDIA') {
            $('.graphics-nvidia').show()
            $('.graphics-unknown').hide()
            $('#graphics-open-driver-name').html('nouveau');
            $('#graphics-proprietary').show();

        } else if (graphicsVendor == "AMD") {
            $('.graphics-amd').show()
            $('.graphics-unknown').hide()
            $('#graphics-open-driver-name').html('radeon');
            //~ $('#graphics-proprietary').show();

        } else if (graphicsVendor == "Intel") {
            $('.graphics-intel').show()
            $('.graphics-unknown').hide()

        } else if (graphicsVendor == "VirtualBox") {
            $('.graphics-vbox').show()
            $('.graphics-unknown').hide()

        } else {
            // Obscure graphics chip or something we can't tell.
            $('#graphics-proprietary').show();
        }
    }, 1000);

    // Expand / Collapse sub-sections to keep it tidy.
    function toggleSub(divID, arrowID) {
        if ($('#' + divID).is(":visible")) {
            $('#' + divID).fadeOut();
            $('#' + arrowID).removeClass('fa-chevron-up');
            $('#' + arrowID).addClass('fa-chevron-down');
        } else {
            $('#' + divID).fadeIn();
            $('#' + arrowID).removeClass('fa-chevron-down');
            $('#' + arrowID).addClass('fa-chevron-up');
        }
    }

    // Fetch system specifications if not cached already.
    // Wait a couple of seconds so it doesn't look like application had frozen.
    function InitSystemInfo() {
        setCursorBusy()
        setTimeout(function () {
            cmd("init-system-info");
        }, 1000);
    }

    // Show popovers on hover.
    $(document).ready(function () {
        $("body").tooltip({
            selector: '[data-toggle=tooltip]'
        });
    });
    $('[rel=unitsinfo]').popover({
        html: true,
        content: function () {
            return $('#popover_units').html();
        }
    });

    // If menu is not open, open menu otherwise goto home page
    function backAction() {
        if (!$('#index-menu').is(':visible')) {
            indexOpen();
        } else {
            if ($('#first-run').is(':visible')) {
                smoothPageFade('default.html');
            } else {
                smoothPageFade('index.html');
            }
        }
    }
}


function sortList() {
    $('#recommendation-list').append($('#recommendation-list > div').sort(function (itemOne, itemTwo) {
        var statusOne = $(itemOne).find('[id$=install]').is(':visible');
        var statusTwo = $(itemTwo).find('[id$=install]').is(':visible');

        if (statusOne && !statusTwo) {
            return -1;
        }

        if (!statusOne && statusTwo) {
            return 1;
        }

        var titleOne = $(itemOne).find('h2').text();
        var titleTwo = $(itemTwo).find('h2').text();
        return titleOne.localeCompare(titleTwo);
    }));
}

function sortAppletList() {
    $('#applet-list').append($('#applet-list > div').sort(function (itemOne, itemTwo) {
        var titleOne = $(itemOne).find('h2').text();
        var titleTwo = $(itemTwo).find('h2').text();
        return titleOne.localeCompare(titleTwo);
    }));
}

function sortThemeList() {
    $('#theme-list').append($('#theme-list > div').sort(function (itemOne, itemTwo) {
        var titleOne = $(itemOne).find('h3').text();
        var titleTwo = $(itemTwo).find('h3').text();
        return titleOne.localeCompare(titleTwo);
    }));
}

function injectTitle() {
    var title = $("#title-inject").text();
    document.title = "title?" + title;
}

injectTitle();